### ISSUE TITLE
if-elsed function definition: Function declaration is obscured by a declaration of the same name

### ISSUE THREAD

**Issue by nschloe** [microsoft/pyright#1871]
Running pyright on
```python
a = 5

if a > 3:  # some expensive test
    def myprint():
        print("one")

else:
    def myprint():
        print("two")


myprint()
```
incorrectly gives
```
[Pyright reportGeneralTypeIssues] [E] Function declaration "myprint" is
               obscured by a declaration of the same name
```
for the first `def myprint()`.

**Comment by erictraut** [comment]
This is intended behavior. If you enable type checking, a symbol cannot be given different type declarations within the same scope. A `def` statement is considered a type declaration.

**Comment by nschloe** [comment]
Interesting. Is this perhaps an alternative design pattern?

**Comment by erictraut** [comment]
I don't understand your question. Can you clarify?

**Comment by nschloe** [comment]
I mean that, apparently the code in the original post is undesirable. Is there are different way to achieve the result that conforms with pyright?

**Comment by erictraut** [comment]
```python
a = 5

def print_one():
    print("one")

def print_two():
    print("two")

myprint = print_one if a > 3 else print_two

myprint()
```

**Comment by nschloe** [comment]
Thanks!

**Comment by carlosgmartin** [comment]
@erictraut Is there a fundamental type-theoretic obstruction that prevents unifying types in the third case below, but not the first two cases?

```python3
if 1:
    a = 3
else:
    a = 4
reveal_type(a)

if 1:
    b = lambda: 3
else:
    b = lambda: 4
reveal_type(b)

if 1:
    def c():
        return 3
else:
    def c():
        return 4
reveal_type(c)
```

```
test.py
  test.py:5:13 - information: Type of "a" is "Literal[3, 4]"
  test.py:11:13 - information: Type of "b" is "(() -> Literal[3]) | (() -> Literal[4])"
  test.py:14:9 - error: Function declaration "c" is obscured by a declaration of the same name (reportGeneralTypeIssues)
  test.py:19:13 - information: Type of "c" is "(() -> Literal[3]) | (() -> Literal[4])"
1 error, 0 warnings, 3 informations
```

You mention

> A `def` statement is considered a type declaration.

Couldn't the type declarations be unified? Couldn't @timarmate-img's solution be applied automatically? It would be nice to have this feature.

**Comment by carlosgmartin** [comment]
Also, it would be nice if the error message were more descriptive, e.g. pointing out the difference in types.

**Comment by erictraut** [comment]
> Is there a fundamental type-theoretic obstruction that prevents unifying types

There isn't a type-theoretic obstruction. the problem is that a `def` statement is considered a "type declaration". In the case of symbols `a` and `b`, you are providing no type declaration, so their types are not constrained or enforced. In the case of `c`, you're providing two contradictory type declarations. It's as if you had done the following:

```python
if 1:
    a: Literal[3] = 3
else:
    a: Literal[4] = 4
```

Contradictory type declarations like this should be flagged as an error by a type checker.

For more information about type declarations, refer to [this documentation](https://microsoft.github.io/pyright/#/type-concepts).

If you change your two declarations of `c` so they are identical, the error will go away.

```python
if 1:
    def c() -> int:
        return 3
else:
    def c() -> int:
        return 4
```

**Comment by carlosgmartin** [comment]
@erictraut Thanks. Is there a fundamental reason why `def` is treated as a type declaration, in contrast to lambda? From an external usability or "semantic correctness" perspective (i.e. putting aside pyright's internal way of doing things), it seems to me like a type checker should accept the (unannotated) examples above.

**Comment by erictraut** [comment]
> Is there a fundamental reason why def is treated as a type declaration, in contrast to lambda?

That's a good question. There are a few reasons why lambdas are not treated as declarations but `def` statements are. 

First, lambdas are expressions, but `def` statements are not. That's an important distinction from a grammar standpoint. There are no other expression forms that are considered declarations, so it would be odd for a lambda expression to be treated as such.

Second, the grammar for lambdas (at least in Python) doesn't support type annotations, so all types associated with a lambda need to be inferred from its usage. This means that type evaluation of lambdas is subject to inference behavior, which is not standardized and varies by type checker. By contrast, `def` statements support full type annotations and are governed by the Python type specification. Their meaning is unequivocal, and they can (and I would argue, should) be interpreted as a clear declaration of the developer's intent.

To be fair, the current draft of the official type spec indicates that a type checker should [not infer return types](https://typing.readthedocs.io/en/latest/spec/annotations.html#the-meaning-of-annotations) for `def` statements — that it should treat all unannotated return types as `Any`. Pyright admittedly violates the spec in this regard because it infers return types. I personally think that's beneficial, but some may disagree. I plan to make the case that the spec should be amended to allow for return type inference. Regardless of whether the spec is amended, it's unlikely that we will change this behavior in pyright. The majority of pyright users like the fact that it infers return types. You are always able to override its inference logic by providing explicit return type annotations, as I demonstrated above. This leaves you in the driver's seat.

**Comment by carlosgmartin** [comment]
@erictraut Thanks again. In that case, perhaps it might be helpful to add more details to the error message, e.g. what the differences are between the conflicting type declarations.
